### ISSUE TITLE
performance: hits dataset performance degradation in new expression

### ISSUE THREAD

**Issue by BohuTANG** [datafuselabs/databend#9483]
**Summary**

The hits table is created and loaded data with the nightly-v0.8.177:

The benchmark with the same settings:
```
mysql> show settings like '%max%';
+------------------------------------+-------------+-------------+---------+-------------------------------------------------------------------------------------------------------------------+--------+
| name                               | value       | default     | level   | description                                                                                                       | type   |
+------------------------------------+-------------+-------------+---------+-------------------------------------------------------------------------------------------------------------------+--------+
| max_block_size                     | 65536       | 65536       | SESSION | Maximum block size for reading, default value: 65536.                                                             | UInt64 |
| max_execute_time                   | 0           | 0           | SESSION | The maximum query execution time. it means no limit if the value is zero. default value: 0.                       | UInt64 |
| max_inlist_to_or                   | 3           | 3           | SESSION | Max size in inlist expression that will convert to or combinator, default value: 3.                               | UInt64 |
| max_memory_usage                   | 26604358860 | 26604358860 | SESSION | The maximum memory usage for processing single query, in bytes. By default the value is determined automatically. | UInt64 |
| max_result_rows                    | 0           | 0           | SESSION | Auto limit max result rows if user not specify the limit, default is 0 means no limit                             | UInt64 |
| max_storage_io_requests            | 64          | 64          | SESSION | The maximum number of concurrent IO requests. By default the value is determined automatically.                   | UInt64 |
| max_threads                        | 32          | 32          | SESSION | The maximum number of threads to execute the request. By default the value is determined automatically.           | UInt64 |
| storage_io_max_page_bytes_for_read | 524288      | 524288      | SESSION | The maximum bytes of one IO request to read. Default the value is 512KB                                           | UInt64 |
+------------------------------------+-------------+-------------+---------+-------------------------------------------------------------------------------------------------------------------+--------+
```

Bench(3rd(Hot), chart generated by [google sheet](https://docs.google.com/spreadsheets/d/1ObfMrknUbq3tgPdVMjlffjVPlF_ARzMPJGtRj-turXs/edit?usp=sharing)):

Update(After #9500 merged)
![image](https://user-images.githubusercontent.com/172204/211149841-948dd7bc-62b1-47f5-a883-80c736308a7b.png)

bench script:
```
cat run-expr.sh
#!/bin/bash

TRIES=3
QUERY_NUM=1

header="| query  | 1st(Cold) | 2nd(Hot) | 3rd(Hot) |\n| ------- | ------- | ------- | ------- |\n";
printf "$header"
cat queries.sql | while read query; do
    echo -n "| q${QUERY_NUM} | "
    for i in $(seq 1 $TRIES); do
        RES=$(curl -w 'Time: %{time_total}\n' http://default@localhost:8124?database=hits -d "${query}" 2>&1 | grep -P '^Time: ' | sed 's/Time: //')
        [[ "$?" == "0" ]] && echo -n "${RES}" || echo -n "null"
        [[ "$i" != $TRIES ]] && echo -n " | "

        echo "${QUERY_NUM},${i},${RES}" >> result.csv
    done
    echo "|"

    QUERY_NUM=$((QUERY_NUM + 1))
done
```

queries.sql is: https://github.com/ClickHouse/ClickBench/blob/main/databend/queries.sql

**Comment by sundy-li** [comment]
We don't respect the old bloom filter index.

**Comment by BohuTANG** [comment]
Update:
![image](https://user-images.githubusercontent.com/172204/210964956-2d18cc08-12c6-4d75-9a6c-971762d02097.png)

Only q14 has some issue:
```
SELECT SearchPhrase, COUNT(DISTINCT UserID) AS u FROM hits WHERE SearchPhrase <> '' GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;
```

**Comment by BohuTANG** [comment]
Update:
new expr explains:
```
mysql> explain SELECT SearchPhrase, COUNT(DISTINCT UserID) AS u FROM hits_expr WHERE SearchPhrase <> '' GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;
+---------------------------------------------------------------------------------------------------+
| explain                                                                                           |
+---------------------------------------------------------------------------------------------------+
| Limit                                                                                             |
| ├── limit: 10                                                                                     |
| ├── offset: 0                                                                                     |
| └── Sort                                                                                          |
|     ├── sort keys: [u DESC NULLS LAST]                                                            |
|     └── EvalScalar                                                                                |
|         ├── expressions: [COUNT(DISTINCT UserID) (#107)]                                          |
|         └── AggregateFinal                                                                        |
|             ├── group by: [searchphrase]                                                          |
|             ├── aggregate functions: [count_distinct(userid)]                                     |
|             └── AggregatePartial                                                                  |
|                 ├── group by: [searchphrase]                                                      |
|                 ├── aggregate functions: [count_distinct(userid)]                                 |
|                 └── Filter                                                                        |
|                     ├── filters: [noteq(hits_expr.searchphrase (#39), "")]                        |
|                     └── TableScan                                                                 |
|                         ├── table: default.hits.hits_expr                                         |
|                         ├── read rows: 99997497                                                   |
|                         ├── read bytes: 569908151                                                 |
|                         ├── partitions total: 751                                                 |
|                         ├── partitions scanned: 751                                               |
|                         ├── push downs: [filters: [not(eq(searchphrase, ""))], limit: NONE]       |
|                         └── output columns: [9, 39]                                               |
+---------------------------------------------------------------------------------------------------+
23 rows in set (0.03 sec)
Read 0 rows, 0.00 B in 0.024 sec., 0 rows/sec., 0.00 B/sec.
```

nightly-v0.8.177 explains:
```
mysql> explain SELECT SearchPhrase, COUNT(DISTINCT UserID) AS u FROM hits WHERE SearchPhrase <> '' GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;
+----------------------------------------------------------------------------------------------+
| explain                                                                                      |
+----------------------------------------------------------------------------------------------+
| Limit                                                                                        |
| ├── limit: 10                                                                                |
| ├── offset: 0                                                                                |
| └── Sort                                                                                     |
|     ├── sort keys: [u DESC NULLS LAST]                                                       |
|     └── EvalScalar                                                                           |
|         ├── expressions: [COUNT(DISTINCT UserID) (#107)]                                     |
|         └── AggregateFinal                                                                   |
|             ├── group by: [searchphrase]                                                     |
|             ├── aggregate functions: [count_distinct(userid)]                                |
|             └── AggregatePartial                                                             |
|                 ├── group by: [searchphrase]                                                 |
|                 ├── aggregate functions: [count_distinct(userid)]                            |
|                 └── Filter                                                                   |
|                     ├── filters: [<>(hits.searchphrase (#39), )]                             |
|                     └── TableScan                                                            |
|                         ├── table: default.hits.hits                                         |
|                         ├── read rows: 99997497                                              |
|                         ├── read bytes: 568828804                                            |
|                         ├── partitions total: 751                                            |
|                         ├── partitions scanned: 751                                          |
|                         ├── push downs: [filters: [<>(searchphrase, '')], limit: NONE]       |
|                         └── output columns: [9, 39]                                          |
+----------------------------------------------------------------------------------------------+
23 rows in set (0.03 sec)
Read 0 rows, 0.00 B in 0.011 sec., 0 rows/sec., 0.00 B/sec.
```

**Comment by xudong963** [comment]
Looks plans are same.

**Comment by BohuTANG** [comment]
new expr q14 pprof:
![image](https://user-images.githubusercontent.com/172204/210970924-081fee57-6b50-45f5-a323-cc637261106e.png)

**Comment by sundy-li** [comment]
I'm working on fix q14.

**Comment by BohuTANG** [comment]
Update(after #9488):
![image](https://user-images.githubusercontent.com/172204/211016973-4693b4f4-73f9-4197-8e43-fe119663977f.png)

cc @sundy-li @xudong963

**Comment by BohuTANG** [comment]
Q40:

## main
```
mysql> explain SELECT TraficSourceID, SearchEngineID, AdvEngineID, CASE WHEN (SearchEngineID = 0 AND AdvEngineID = 0) THEN Referer ELSE '' END AS Src, URL AS Dst, COUNT(*) AS PageViews FROM hits_expr WHERE CounterID = 62 AND EventDate >= '2013-07-01' AND EventDate <= '2013-07-31' AND IsRefresh = 0 GROUP BY TraficSourceID, SearchEngineID, AdvEngineID, Src, Dst ORDER BY PageViews DESC LIMIT 10 OFFSET 1000;
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                                                                                                             |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EvalScalar                                                                                                                                                                                                          |
| ├── expressions: [group_item (#107)]                                                                                                                                                                                |
| └── Limit                                                                                                                                                                                                           |
|     ├── limit: 10                                                                                                                                                                                                   |
|     ├── offset: 1000                                                                                                                                                                                                |
|     └── Sort                                                                                                                                                                                                        |
|         ├── sort keys: [pageviews DESC NULLS LAST]                                                                                                                                                                  |
|         └── EvalScalar                                                                                                                                                                                              |
|             ├── expressions: [COUNT(*) (#109)]                                                                                                                                                                      |
|             └── AggregateFinal                                                                                                                                                                                      |
|                 ├── group by: [traficsourceid, searchengineid, advengineid, src, url]                                                                                                                               |
|                 ├── aggregate functions: [count()]                                                                                                                                                                  |
|                 └── AggregatePartial                                                                                                                                                                                |
|                     ├── group by: [traficsourceid, searchengineid, advengineid, src, url]                                                                                                                           |
|                     ├── aggregate functions: [count()]                                                                                                                                                              |
|                     └── EvalScalar                                                                                                                                                                                  |
|                         ├── expressions: [if(and(eq(hits_expr.searchengineid (#38), 0_u8), eq(hits_expr.advengineid (#40), 0_u8)), hits_expr.referer (#14), "")]                                                    |
|                         └── Filter                                                                                                                                                                                  |
|                             ├── filters: [eq(hits_expr.counterid (#6), 62_u8), gte(hits_expr.eventdate (#5), "2013-07-01"), lte(hits_expr.eventdate (#5), "2013-07-31"), eq(hits_expr.isrefresh (#15), 0_u8)]       |
|                             └── TableScan                                                                                                                                                                           |
|                                 ├── table: default.hits.hits_expr                                                                                                                                                   |
|                                 ├── read rows: 6720676                                                                                                                                                              |
|                                 ├── read bytes: 403467467                                                                                                                                                           |
|                                 ├── partitions total: 751                                                                                                                                                           |
|                                 ├── partitions scanned: 52                                                                                                                                                          |
|                                 ├── push downs: [filters: [eq(counterid, 62_i32), not(lt(eventdate, 2013-07-01)), not(gt(eventdate, 2013-07-31)), eq(isrefresh, 0_i16)], limit: NONE]                               |
|                                 └── output columns: [5, 6, 13, 14, 15, 37, 38, 40]                                                                                                                                  |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
## nightly-v0.8.177
```
mysql> explain SELECT TraficSourceID, SearchEngineID, AdvEngineID, CASE WHEN (SearchEngineID = 0 AND AdvEngineID = 0) THEN Referer ELSE '' END AS Src, URL AS Dst, COUNT(*) AS PageViews FROM hits WHERE CounterID = 62 AND EventDate >= '2013-07-01' AND EventDate <= '2013-07-31' AND IsRefresh = 0 GROUP BY TraficSourceID, SearchEngineID, AdvEngineID, Src, Dst ORDER BY PageViews DESC LIMIT 10 OFFSET 1000;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                                                                           |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Limit                                                                                                                                                                             |
| ├── limit: 10                                                                                                                                                                     |
| ├── offset: 1000                                                                                                                                                                  |
| └── EvalScalar                                                                                                                                                                    |
|     ├── expressions: [group_item (#107)]                                                                                                                                          |
|     └── Sort                                                                                                                                                                      |
|         ├── sort keys: [pageviews DESC NULLS LAST]                                                                                                                                |
|         └── EvalScalar                                                                                                                                                            |
|             ├── expressions: [COUNT(*) (#109)]                                                                                                                                    |
|             └── AggregateFinal                                                                                                                                                    |
|                 ├── group by: [traficsourceid, searchengineid, advengineid, src, url]                                                                                             |
|                 ├── aggregate functions: [count()]                                                                                                                                |
|                 └── AggregatePartial                                                                                                                                              |
|                     ├── group by: [traficsourceid, searchengineid, advengineid, src, url]                                                                                         |
|                     ├── aggregate functions: [count()]                                                                                                                            |
|                     └── EvalScalar                                                                                                                                                |
|                         ├── expressions: [if(and(=(hits.searchengineid (#38), 0), =(hits.advengineid (#40), 0)), hits.referer (#14), )]                                           |
|                         └── Filter                                                                                                                                                |
|                             ├── filters: [=(hits.counterid (#6), 62), >=(hits.eventdate (#5), 2013-07-01), <=(hits.eventdate (#5), 2013-07-31), =(hits.isrefresh (#15), 0)]       |
|                             └── TableScan                                                                                                                                         |
|                                 ├── table: default.hits.hits                                                                                                                      |
|                                 ├── read rows: 6574090                                                                                                                            |
|                                 ├── read bytes: 397055700                                                                                                                         |
|                                 ├── partitions total: 751                                                                                                                         |
|                                 ├── partitions scanned: 51                                                                                                                        |
|                                 ├── push downs: [filters: [(counterid = 62), (eventdate >= '2013-07-01'), (eventdate <= '2013-07-31'), (isrefresh = 0)], limit: NONE]             |
|                                 └── output columns: [5, 6, 13, 14, 15, 37, 38, 40]                                                                                                |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
27 rows in set (0.04 sec)
Read 0 rows, 0.00 B in 0.025 sec., 0 rows/sec., 0.00 B/sec.
```

**Comment by BohuTANG** [comment]
Update(0107):
![image](https://user-images.githubusercontent.com/172204/211135196-f051462a-df26-461b-ae84-1867b37207bd.png)

**Comment by BohuTANG** [comment]
Update(After #9500 merged)
![image](https://user-images.githubusercontent.com/172204/211149841-948dd7bc-62b1-47f5-a883-80c736308a7b.png)

cc @sundy-li

**Comment by BohuTANG** [comment]
![image](https://user-images.githubusercontent.com/172204/211196543-4aee28d2-18a7-4615-911f-af361f423f3e.png)
cc @sundy-li

**Comment by sundy-li** [comment]
@BohuTANG  can we close it? It's faster now!

**Comment by BohuTANG** [comment]
![image](https://user-images.githubusercontent.com/172204/211795330-b3778584-8b8a-416f-b6f4-95ea1b7e1384.png)

cc @sundy-li

**Comment by BohuTANG** [comment]
Q37&Q38 still have some issues(test them twice the result are same), I prefer to close this and file a new issue (If we address them someday).
